### PR TITLE
Add CNN-DM dataset to torchtext

### PR DESCRIPTION
**Pull request by pmabbo13**
# Description
Add CNNDM dataset to TorchText using TorchData datapipes

# Process

1. Download CNN and DailyMail data from [here](https://cs.nyu.edu/~kcho/DMQA/). These downloads are cached, however the extractions from the tar files are not yet cached. See follow-up items. 
2. Download URL lists from [here](https://github.com/becxer/cnn-dailymail/), which designate which stories belong to the train/dev/test splits. These download are not yet cached. See follow-up items.
3. Use URL list to filter for stories that belong to target split.
4. Parse these stories to separate the article text from the abstract text and return a datapipe that yields (article, abstract) for each story.

# Testing
Create mock dataset that mimics the format of raw CNNDM files and ensure that the datapipe yields the correct output on this mock dataset

`pytest test/datasets/test_cnndm.py`

# Follow-Up Items

- [ ] Cache the extracted tar files for cnn_stories.tgz and dailymail_stories.tgz
- [ ] Cache URL lists (either on disk or using LRUCache) pending benchmark tests on how long it takes to instantiate them on the fly vs loading from cache

### CODE REVIEW

File: torchtext/data/datasets_utils.py
Comment (Nayef211) [inline torchtext/data/datasets_utils.py:403]: nit: remove new line

File: torchtext/data/datasets_utils.py
Comment (Nayef211) [inline torchtext/data/datasets_utils.py:353]: nit: remove new line

File: torchtext/datasets/cnndm.py
Comment (Nayef211) [inline torchtext/datasets/cnndm.py:31]: Can we delete this if we're not planning to use it anywhere?

File: torchtext/data/datasets_utils.py
Comment (Nayef211) [inline torchtext/data/datasets_utils.py:316]: For both of the `parse_cnndm` datapipes, let's mention in docstrings that the code is inspired from https://github.com/abisee/cnn-dailymail/blob/master/make_datafiles.py

File: torchtext/datasets/cnndm.py
Comment (Nayef211) [inline torchtext/datasets/cnndm.py:64]: nit: remove new line

File: torchtext/datasets/cnndm.py
Comment (Nayef211) [inline torchtext/datasets/cnndm.py:55]: leave a comment here saying that this function will be used to cache the contents of the tar file

**Review by Nayef211 — COMMENTED** [review]
Overall the implementation looks great! Some followup items:
- Update the PR description
- Add a section in the description for followup items with a checklist of features that have yet to be implemented. We can then add a link to followup PRs that implement those features and check them off (similar to https://github.com/pytorch/text/pull/1542)
- Add unittests for the dataset

**Review by Nayef211 — APPROVED** [review]
LGTM! Thanks for addressing all the PR comments. Let's wait for a review from @parmeet before we merge this in. 😄

File: torchtext/datasets/cnndm.py
Comment (parmeet) [inline torchtext/datasets/cnndm.py:61]: `parse_cnndm_split` is essentially mapping URLs to corresponding hash+'.story' right? If so, instead of creating DP, could we just create a private function and then use `map`.

File: test/datasets/test_cnndm.py
Comment (parmeet) [inline test/datasets/test_cnndm.py:42]: nit: I wonder if we should create multiple URLs per story type (CNN, dailymail) like the real datasets? In general I think it's fine what we did here.

**Review by parmeet — APPROVED** [review]

File: torchtext/datasets/cnndm.py
Comment (Nayef211) [reply torchtext/datasets/cnndm.py:61]: Do we have any guidance on when to create a DP to do some processing vs using map + private function? Are we only creating a DP for parsing through the dataset?

Comment (Nayef211) [reply torchtext/datasets/cnndm.py]: I spoke with @parmeet offline and I think the sentiment was that if the processing is simple and is mapping some input text into some output text as is done by `parse_cnndm_split` then we should use the `map` approach. If on the otherhand the processing complex as is the case for `parse_cnndm_data` then we should convert the helper function into a datapipe. Sorry for the extra work to change this back to a private fn!

**Pull request merged**
